File: python/test/t_MeasureEvaluation_edge.py
#!/usr/bin/env python

import openturns as ot
import openturns.testing as ott
import otrobopt
import os
import tempfile

# Test IndividualChanceMeasure with discrete distribution
# (exercises the branch where weights index was wrong)
ot.RandomGenerator.SetSeed(0)

f_base = ot.SymbolicFunction(["x", "theta"], ["x*theta"])
f = ot.ParametricFunction(f_base, [1], [1.0])

x = [2.0]

# Discrete distribution with 5 equally-weighted points
discrete_points = [[0.0], [0.5], [1.0], [1.5], [2.0]]
probas = [0.2] * 5
discrete_dist = ot.UserDefined(discrete_points, probas)

icm = otrobopt.IndividualChanceMeasure(f, discrete_dist, ot.Greater(), [0.5])
val = icm(x)
ott.assert_almost_equal(val, [8.12589], 1e-4, 1e-4)

# JointChanceMeasure with discrete distribution
jcm = otrobopt.JointChanceMeasure(f, discrete_dist, ot.Greater(), 0.9)
val2 = jcm(x)
ott.assert_almost_equal(val2, [6.84434], 1e-4, 1e-4)

# QuantileMeasure with discrete distribution
qm = otrobopt.QuantileMeasure(f, discrete_dist, 0.5)
val3 = qm(x)
ott.assert_almost_equal(val3, [2.0], 1e-4, 1e-4)

# MeanStandardDeviationTradeoffMeasure with discrete distribution
mstd = otrobopt.MeanStandardDeviationTradeoffMeasure(f, discrete_dist, [0.8])
val4 = mstd(x)
ott.assert_almost_equal(val4, [1.53137], 1e-4, 1e-4)

# VarianceMeasure with discrete distribution
vm = otrobopt.VarianceMeasure(f, discrete_dist)
val5 = vm(x)
ott.assert_almost_equal(val5, [2.0], 1e-4, 1e-4)

# AggregatedMeasure with discrete distributions
measures = otrobopt.AggregatedMeasure(
    [otrobopt.MeanMeasure(f, discrete_dist), otrobopt.VarianceMeasure(f, discrete_dist)]
)
val6 = measures(x)
ott.assert_almost_equal(val6, [2.0, 2.0], 1e-4, 1e-4)

# Multi-output IndividualChanceMeasure with discrete distribution
f_base2 = ot.SymbolicFunction(["x", "theta"], ["x*theta", "x+theta"])
f2 = ot.ParametricFunction(f_base2, [1], [1.0])
icm2 = otrobopt.IndividualChanceMeasure(f2, discrete_dist, ot.Greater(), [0.5, 0.5])
val7 = icm2(x)
ott.assert_almost_equal(val7.getDimension(), 2)

# Test save/load of measures with discrete distributions
filename = tempfile.NamedTemporaryFile(suffix=".xml", delete=False).name
try:
    study = ot.Study()
    study.setStorageManager(ot.XMLStorageManager(filename))
    study.add("icm", icm)
    study.add("jcm", jcm)
    study.save()

    study2 = ot.Study()
    study2.setStorageManager(ot.XMLStorageManager(filename))
    study2.load()
    icm_loaded = otrobopt.IndividualChanceMeasure()
    study2.fillObject("icm", icm_loaded)
    ott.assert_almost_equal(icm_loaded(x), val, 1e-4, 1e-4)
finally:
    if os.path.exists(filename):
        os.remove(filename)

File: lib/test/t_MeasureEvaluation_edge.cxx
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"
#include "otrobopt/OTRobOpt.hxx"

using namespace OT;
using namespace OT::Test;
using namespace OTROBOPT;

int main()
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  try
  {
    // ---------------------------------------------------------------
    // 1) IndividualChanceMeasure with discrete distribution
    //    (exercises the discrete branch where weights index was wrong)
    // ---------------------------------------------------------------
    {
      Description input({"x", "theta"});
      SymbolicFunction f_base(input, Description({"x*theta"}));
      ParametricFunction f(f_base, Indices(1, 1), Point(1, 1.0));

      Point x(1, 2.0);

      // FiniteDiscreteDistribution discrete distribution with 5 points
      Sample discretePoints(5, 1);
      discretePoints(0, 0) = 0.0;
      discretePoints(1, 0) = 0.5;
      discretePoints(2, 0) = 1.0;
      discretePoints(3, 0) = 1.5;
      discretePoints(4, 0) = 2.0;
      Point probas(5, 0.2);
      FiniteDiscreteDistribution discreteDist(discretePoints, probas);

      IndividualChanceMeasure icm(f, discreteDist, Greater(), Point(1, 0.5));
      Point val(icm(x));
      fullprint << "IndividualChance (discrete)=" << val << std::endl;
      // theta values >= 0: all 5 points have f=2*theta >=0 => prob=1.0
      // Greater operator: result = P(f>=0) - alpha = 1.0 - 0.5 = 0.5
      assert_almost_equal(val, Point(1, 0.5), 1e-4, 1e-4);
    }

    // ---------------------------------------------------------------
    // 2) IndividualChanceMeasure with discrete distribution, multi-output
    // ---------------------------------------------------------------
    {
      Description input({"x", "theta"});
      SymbolicFunction f_base(input, Description({"x*theta", "x+theta"}));
      ParametricFunction f(f_base, Indices(1, 1), Point(1, 1.0));

      Point x(1, 1.0);
      Sample discretePoints(3, 1);
      discretePoints(0, 0) = -1.0;
      discretePoints(1, 0) = 0.0;
      discretePoints(2, 0) = 1.0;
      Point probas(3, 1.0 / 3.0);
      FiniteDiscreteDistribution discreteDist(discretePoints, probas);

      IndividualChanceMeasure icm(f, discreteDist, Greater(), Point(2, 0.5));
      Point val(icm(x));
      fullprint << "IndividualChance (discrete, multi-output)=" << val << std::endl;
      assert_almost_equal(val.getDimension(), 2);
    }

    // ---------------------------------------------------------------
    // 3) JointChanceMeasure with discrete distribution
    // ---------------------------------------------------------------
    {
      Description input({"x", "theta"});
      SymbolicFunction f_base(input, Description({"x*theta"}));
      ParametricFunction f(f_base, Indices(1, 1), Point(1, 1.0));

      Point x(1, 1.0);
      Sample discretePoints(3, 1);
      discretePoints(0, 0) = -1.0;
      discretePoints(1, 0) = 0.0;
      discretePoints(2, 0) = 1.0;
      Point probas(3, 1.0 / 3.0);
      FiniteDiscreteDistribution discreteDist(discretePoints, probas);

      JointChanceMeasure jcm(f, discreteDist, Greater(), 0.9);
      Point val(jcm(x));
      fullprint << "JointChance (discrete)=" << val << std::endl;
      assert_almost_equal(val.getDimension(), 1);
    }

    // ---------------------------------------------------------------
    // 4) QuantileMeasure with discrete distribution
    // ---------------------------------------------------------------
    {
      Description input({"x", "theta"});
      SymbolicFunction f_base(input, Description({"x*theta"}));
      ParametricFunction f(f_base, Indices(1, 1), Point(1, 1.0));

      Point x(1, 1.0);
      Sample discretePoints(3, 1);
      discretePoints(0, 0) = 0.0;
      discretePoints(1, 0) = 1.0;
      discretePoints(2, 0) = 2.0;
      Point probas(3, 1.0 / 3.0);
      FiniteDiscreteDistribution discreteDist(discretePoints, probas);

      QuantileMeasure qm(f, discreteDist, 0.5);
      Point val(qm(x));
      fullprint << "Quantile (discrete)=" << val << std::endl;
      assert_almost_equal(val.getDimension(), 1);
    }

    // ---------------------------------------------------------------
    // 5) MeanStandardDeviationTradeoffMeasure with discrete distribution
    // ---------------------------------------------------------------
    {
      Description input({"x", "theta"});
      SymbolicFunction f_base(input, Description({"x*theta"}));
      ParametricFunction f(f_base, Indices(1, 1), Point(1, 1.0));

      Point x(1, 1.0);
      Sample discretePoints(3, 1);
      discretePoints(0, 0) = 0.0;
      discretePoints(1, 0) = 1.0;
      discretePoints(2, 0) = 2.0;
      Point probas(3, 1.0 / 3.0);
      FiniteDiscreteDistribution discreteDist(discretePoints, probas);

      MeanStandardDeviationTradeoffMeasure mstd(f, discreteDist, Point(1, 0.5));
      Point val(mstd(x));
      fullprint << "MeanStdTradeoff (discrete)=" << val << std::endl;
      assert_almost_equal(val.getDimension(), 1);
    }

    // ---------------------------------------------------------------
    // 6) VarianceMeasure with discrete distribution
    // ---------------------------------------------------------------
    {
      Description input({"x", "theta"});
      SymbolicFunction f_base(input, Description({"x*theta"}));
      ParametricFunction f(f_base, Indices(1, 1), Point(1, 1.0));

      Point x(1, 1.0);
      Sample discretePoints(3, 1);
      discretePoints(0, 0) = 0.0;
      discretePoints(1, 0) = 1.0;
      discretePoints(2, 0) = 2.0;
      Point probas(3, 1.0 / 3.0);
      FiniteDiscreteDistribution discreteDist(discretePoints, probas);

      VarianceMeasure vm(f, discreteDist);
      Point val(vm(x));
      fullprint << "Variance (discrete)=" << val << std::endl;
      assert_almost_equal(val.getDimension(), 1);
    }

    // ---------------------------------------------------------------
    // 7) AggregatedMeasure with discrete distributions
    // ---------------------------------------------------------------
    {
      Description input({"x", "theta"});
      SymbolicFunction f_base(input, Description({"x*theta"}));
      ParametricFunction f(f_base, Indices(1, 1), Point(1, 1.0));

      Point x(1, 1.0);
      Sample discretePoints(3, 1);
      discretePoints(0, 0) = 0.0;
      discretePoints(1, 0) = 1.0;
      discretePoints(2, 0) = 2.0;
      Point probas(3, 1.0 / 3.0);
      FiniteDiscreteDistribution discreteDist(discretePoints, probas);

      Collection<MeasureEvaluation> measures;
      measures.add(MeanMeasure(f, discreteDist));
      measures.add(VarianceMeasure(f, discreteDist));

      AggregatedMeasure agg(measures);
      Point val(agg(x));
      fullprint << "Aggregated (discrete)=" << val << std::endl;
      assert_almost_equal(val.getDimension(), 2);
    }

    // ---------------------------------------------------------------
    // 8) MeasureEvaluation copy-on-write behavior
    // ---------------------------------------------------------------
    {
      Description input({"x", "theta"});
      SymbolicFunction f_base(input, Description({"x*theta"}));
      ParametricFunction f(f_base, Indices(1, 1), Point(1, 1.0));
      Normal thetaDist(2.0, 0.1);

      MeasureEvaluation m1(MeanMeasure(f, thetaDist));
      MeasureEvaluation m2(m1);
      // Both share the same implementation; m1 and m2 evaluate identically
      Point x(1, 1.0);
      assert_almost_equal(m1(x), m2(x));
    }
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }
  return ExitCode::Success;
}
